ui-admin/src/forms/designer/questions/ChoicesList.tsx

```diff
@@ -0,0 +1,158 @@
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
+import { faChevronDown, faChevronUp, faPlus, faTimes } from '@fortawesome/free-solid-svg-icons'
+import React, { useId } from 'react'
+
+import { CheckboxQuestion, DropdownQuestion, RadiogroupQuestion } from '@juniper/ui-core'
+
+import { Button, IconButton } from 'components/forms/Button'
+import { TextInput } from 'components/forms/TextInput'
+
+type QuestionWithChoices = CheckboxQuestion | DropdownQuestion | RadiogroupQuestion
+
+type ChoicesListProps = {
+  question: QuestionWithChoices
+  readOnly: boolean
+  onChange: (newValue: QuestionWithChoices) => void
+}
+
+/** UI for editing the list of choices for a question. */
+export const ChoicesList = (props: ChoicesListProps) => {
+  const { question, readOnly, onChange } = props
+
+  const labelId = useId()
+
+  return (
+    <div className="mb-3">
+      <p className="mb-2" id={labelId}>Choices</p>
+      <ol aria-labelledby={labelId} className="list-group list-group-numbered mb-1">
+        {question.choices.map((choice, i) => {
+          return (
+            <li
+              key={i}
+              className="list-group-item d-flex"
+            >
+              <div
+                className="flex-grow-1 ms-3"
+                style={{
+                  display: 'grid',
+                  grid: 'auto-flow / max-content auto',
+                  gridColumnGap: '0.5rem',
+                  gridRowGap: '1rem',
+                  alignItems: 'center',
+                  height: 'fit-content'
+                }}
+              >
+                <TextInput
+                  className="form-control"
+                  disabled={readOnly}
+                  label="Text"
+                  labelClassname="mb-0"
+                  value={choice.text}
+                  onChange={value => {
+                    onChange({
+                      ...question,
```

Comment:
this pattern makes me nervous -- since you're not deep-copying question, it seems more likely that bugs related to duplicate references being passed across state updates will result if `question` acquires nested attributes.  
This also seems like it's complex enough to warrant being in a named function.  `changeChoiceText(index, value)`

Reply:
I'm curious about your thinking here, but not quite following. I agree that the `...` spread operator can be a bit loose, but it's also very powerful, and using typescript also focuses the range of possible values in a way that makes me less nervous about it.

Reply:
I don't think this component needs to care whether or not something else is mutating other fields. All it does is modify the choices list.

We could make this terser with Lodash/FP, but at the cost of the type safety of built in operators.